add aac support

baycom opened the pull request:
Requirements:
- Fraunhofer FDK AAC library 
- aac capable libshout from https://github.com/codders/libshout

New config parameters:
- aot: determines Audio Object Type (one of "lc", "he", "hev2"), default "lc"
- aacenc_afterburner: This parameter controls the use of the afterburner feature. The after- burner is a type of analysis by synthesis algorithm which increases the audio quality but also the required processing power (one of "no", "yes"), default "yes" 
- quality: Enables VBR (one of "0"..."8", depending on AOT), default 0 (CBR enabled)
- bitrate: Sets CBR, default "128"

MaxKellermann commented:
Something went wrong with your commit - it deletes the dbus/udisks build rules!
Rule of thumb: if a PR needs explanatory text, this means that your commit message is bad.
Please add documentation to doc/user.xml.
Is this libshout you mentioned an unofficial fork? Will this be merged into the real libshout eventually? Is this fork being adopted by distributions?
Your code has no compile-time check for which libshout version is available; this will break the build with AAC-incapable libshout versions.

MaxKellermann commented on Makefile.am:
```diff
@@ -46,7 +46,6 @@ src_mpd_LDADD = \
 	$(STORAGE_LIBS) \
 	$(PLAYLIST_LIBS) \
 	$(AVAHI_LIBS) \
-	$(DBUS_LIBS) \
```
Uh...?

MaxKellermann commented on src/encoder/plugins/FdkaacEncoderPlugin.cxx:
```diff
@@ -0,0 +1,263 @@
+		}else if(!strcmp(value, "eld")) {
+			aot=AOT_ER_AAC_ELD;
+		}else {
+			aot=AOT_AAC_LC;
```
If an invalid value is configured, this should not be silently ignored.

MaxKellermann commented on src/encoder/plugins/FdkaacEncoderPlugin.cxx:
```diff
@@ -0,0 +1,263 @@
+	audio_format.format = SampleFormat::S16;
+	audio_format.channels = 2;
+	AACENC_InfoStruct *info = (AACENC_InfoStruct *)calloc(sizeof(AACENC_InfoStruct), 1);
+	AacEncoder = (HANDLE_AACENCODER*)calloc(sizeof(HANDLE_AACENCODER), 1);
```
Does the library really require zero-initialization? Or what is the justification for using this unsafe plain C function?

MaxKellermann commented on src/encoder/plugins/FdkaacEncoderPlugin.cxx:
```diff
@@ -0,0 +1,263 @@
+	AacEncoder = (HANDLE_AACENCODER*)calloc(sizeof(HANDLE_AACENCODER), 1);
+	AACENC_ERROR res = aacEncOpen ( AacEncoder, 0, 0);
+	if (res != AACENC_OK)
+		throw std::runtime_error("aacEncOpen failed");
```
This leaks memory.

MaxKellermann commented on src/encoder/plugins/FdkaacEncoderPlugin.cxx:
```diff
@@ -0,0 +1,263 @@
+	assert(output_begin == output_end);
+	const unsigned char *src = (const unsigned char *)data;
+
+	int frame_size = info->frameLength*audio_format.GetSampleSize()*audio_format.channels;	// bytes
```
Don't use signed integers for memory sizes! That's what we have `size_t` for.

MaxKellermann commented on src/encoder/plugins/FdkaacEncoderPlugin.cxx:
```diff
@@ -0,0 +1,263 @@
+
+	int frame_size = info->frameLength*audio_format.GetSampleSize()*audio_format.channels;	// bytes
+
+	int read_buffer_pos=0;
```
Same here: no `int`!

MaxKellermann commented on src/encoder/plugins/FdkaacEncoderPlugin.cxx:
```diff
@@ -0,0 +1,263 @@
+
+		size_t bytes_to_copy = frame_size-input_buffer_pos;
+		if(bytes_to_copy > length) {
+			bytes_to_copy = length;	
```
Why not `std::min()`?

MaxKellermann commented on src/encoder/plugins/FdkaacEncoderPlugin.cxx:
```diff
@@ -0,0 +1,263 @@
+
+	ReusableArray<unsigned char, 32768> output_buffer;
+	ReusableArray<unsigned char, 32768> input_buffer;
+	unsigned char *output_begin = nullptr, *output_end = nullptr, *input_begin = nullptr;
```
These attributes should be documented.

MaxKellermann commented on src/encoder/plugins/FdkaacEncoderPlugin.cxx:
```diff
@@ -0,0 +1,263 @@
+	int input_buffer_pos = 0;
+
+	ReusableArray<unsigned char, 32768> output_buffer;
+	ReusableArray<unsigned char, 32768> input_buffer;
```
Why `32768`? Do you know what this template parameter means?

MaxKellermann commented on src/encoder/plugins/FdkaacEncoderPlugin.cxx:
```diff
@@ -0,0 +1,263 @@
+	const AudioFormat audio_format;
+	HANDLE_AACENCODER *AacEncoder;
+	AACENC_InfoStruct *info;
+	int input_buffer_pos = 0;
```
No signed integer! This is asking for trouble.

MaxKellermann commented on src/encoder/plugins/FdkaacEncoderPlugin.cxx:
```diff
@@ -0,0 +1,263 @@
+		src += bytes_to_copy;
+
+		if(input_buffer_pos == frame_size) {
+			const auto dest = output_buffer.Get(info->maxOutBufBytes);
```
Does this overwrite the contents of your output buffer in each loop iteration?

MaxKellermann commented on src/encoder/plugins/FdkaacEncoderPlugin.cxx:
```diff
@@ -0,0 +1,263 @@
+
+			AACENC_ERROR res = aacEncEncode ( *AacEncoder, &in_buf, &out_buf, &in_args, &out_args );
+			if(res != AACENC_OK) {
+				if (res == AACENC_ENCODE_EOF) {
```
Why this redundant comparison?

MaxKellermann commented on src/encoder/plugins/FdkaacEncoderPlugin.cxx:
```diff
@@ -0,0 +1,263 @@
+	assert(begin <= output_end);
+	const size_t remainning = output_end - begin;
+	if (length > remainning)
+		length = remainning;
```
`std::min()`

MaxKellermann commented:
Some research about Fraunhofer FDK AAC library showed that the library is non-free, thus incompatible with MPD's license, and your code would be illegal in MPD. Sorry.

baycom commented:
On 05.06.18 11:09, Max Kellermann wrote:
>
> Something went wrong with your commit - it deletes the dbus/udisks 
> build rules!
>
fixed.

> Rule of thumb: if a PR needs explanatory text, this means that your 
> commit message is bad.
> Please add documentation to doc/user.xml.
>
fixed.
>
> Is this libshout you mentioned an unofficial fork? Will this be merged 
> into the real libshout eventually? Is this fork being adopted by 
> distributions?
> Your code has no compile-time check for which libshout version is 
> available; this will break the build with AAC-incapable libshout versions.
>
There is a pull request that should be triggered again 
(https://github.com/codders/libshout/issues/4)

baycom commented:
On 05.06.18 11:19, Max Kellermann wrote:
>
> *@MaxKellermann* requested changes on this pull request.
>
> ------------------------------------------------------------------------
>
> In Makefile.am 
> <https://github.com/MusicPlayerDaemon/MPD/pull/287#discussion_r192999743>:
>
> > @@ -46,7 +46,6 @@ src_mpd_LDADD = \
>   	$(STORAGE_LIBS) \
>   	$(PLAYLIST_LIBS) \
>   	$(AVAHI_LIBS) \
> -	$(DBUS_LIBS) \
>
> Uh...?
>
Some merge bug, fixed.

> ------------------------------------------------------------------------
>
> In src/encoder/plugins/FdkaacEncoderPlugin.cxx 
> <https://github.com/MusicPlayerDaemon/MPD/pull/287#discussion_r193000303>:
>
> > +	const char *value;
> +
> +	value = block.GetBlockValue("aot", "lc");
> +	if (value != nullptr) {
> +		if(!strcmp(value, "lc")) {
> +			aot=AOT_AAC_LC;
> +		}else if(!strcmp(value, "he")) {
> +			aot=AOT_SBR;
> +		}else if(!strcmp(value, "hev2")) {
> +			aot=AOT_PS;
> +		}else if(!strcmp(value, "ld")) {
> +			aot=AOT_ER_AAC_LD;
> +		}else if(!strcmp(value, "eld")) {
> +			aot=AOT_ER_AAC_ELD;
> +		}else {
> +			aot=AOT_AAC_LC;
>
> If an invalid value is configured, this should not be silently ignored.
>
fixed.
> ------------------------------------------------------------------------
>
> In src/encoder/plugins/FdkaacEncoderPlugin.cxx 
> <https://github.com/MusicPlayerDaemon/MPD/pull/287#discussion_r193000431>:
>
> > +	}
> +	if(aacEncoder_SetParam(*AacEncoder, AACENC_AFTERBURNER, aacenc_afterburner) != AACENC_OK) {
> +		throw std::runtime_error("error setting fdkaac afterburner");
> +	}
> +	if (aacEncEncode(*AacEncoder, NULL, NULL, NULL, NULL) != AACENC_OK) {
> +		throw std::runtime_error("Unable to initialize the encoder\n");
> +	}
> +}
> +
> +Encoder *
> +PreparedFdkaacEncoder::Open(AudioFormat &audio_format)
> +{
> +	audio_format.format = SampleFormat::S16;
> +	audio_format.channels = 2;
> +	AACENC_InfoStruct *info = (AACENC_InfoStruct *)calloc(sizeof(AACENC_InfoStruct), 1);
> +	AacEncoder = (HANDLE_AACENCODER*)calloc(sizeof(HANDLE_AACENCODER), 1);
>
> Does the library really require zero-initialization? Or what is the 
> justification for using this unsafe plain C function?
>
fixed.
> ------------------------------------------------------------------------
>
> In src/encoder/plugins/FdkaacEncoderPlugin.cxx 
> <https://github.com/MusicPlayerDaemon/MPD/pull/287#discussion_r193000481>:
>
> > +	}
> +	if (aacEncEncode(*AacEncoder, NULL, NULL, NULL, NULL) != AACENC_OK) {
> +		throw std::runtime_error("Unable to initialize the encoder\n");
> +	}
> +}
> +
> +Encoder *
> +PreparedFdkaacEncoder::Open(AudioFormat &audio_format)
> +{
> +	audio_format.format = SampleFormat::S16;
> +	audio_format.channels = 2;
> +	AACENC_InfoStruct *info = (AACENC_InfoStruct *)calloc(sizeof(AACENC_InfoStruct), 1);
> +	AacEncoder = (HANDLE_AACENCODER*)calloc(sizeof(HANDLE_AACENCODER), 1);
> +	AACENC_ERROR res = aacEncOpen ( AacEncoder, 0, 0);
> +	if (res != AACENC_OK)
> +		throw std::runtime_error("aacEncOpen failed");
>
> This leaks memory.
>
fixed.
> ------------------------------------------------------------------------
>
> In src/encoder/plugins/FdkaacEncoderPlugin.cxx 
> <https://github.com/MusicPlayerDaemon/MPD/pull/287#discussion_r193000776>:
>
> > +}
> +
> +FdkaacEncoder::~FdkaacEncoder()
> +{
> +	aacEncClose (AacEncoder);
> +	free(AacEncoder);
> +	free(info);
> +}
> +
> +void
> +FdkaacEncoder::Write(const void *data, size_t length)
> +{
> +	assert(output_begin == output_end);
> +	const unsigned char *src = (const unsigned char *)data;
> +
> +	int frame_size = info->frameLength*audio_format.GetSampleSize()*audio_format.channels;	// bytes
>
> Don't use signed integers for memory sizes! That's what we have 
> |size_t| for.
>
fixed.
> ------------------------------------------------------------------------
>
> In src/encoder/plugins/FdkaacEncoderPlugin.cxx 
> <https://github.com/MusicPlayerDaemon/MPD/pull/287#discussion_r193000815>:
>
> > +FdkaacEncoder::~FdkaacEncoder()
> +{
> +	aacEncClose (AacEncoder);
> +	free(AacEncoder);
> +	free(info);
> +}
> +
> +void
> +FdkaacEncoder::Write(const void *data, size_t length)
> +{
> +	assert(output_begin == output_end);
> +	const unsigned char *src = (const unsigned char *)data;
> +
> +	int frame_size = info->frameLength*audio_format.GetSampleSize()*audio_format.channels;	// bytes
> +
> +	int read_buffer_pos=0;
>
> Same here: no |int|!
>
fixed.
> ------------------------------------------------------------------------
>
> In src/encoder/plugins/FdkaacEncoderPlugin.cxx 
> <https://github.com/MusicPlayerDaemon/MPD/pull/287#discussion_r193000891>:
>
> > +}
> +
> +void
> +FdkaacEncoder::Write(const void *data, size_t length)
> +{
> +	assert(output_begin == output_end);
> +	const unsigned char *src = (const unsigned char *)data;
> +
> +	int frame_size = info->frameLength*audio_format.GetSampleSize()*audio_format.channels;	// bytes
> +
> +	int read_buffer_pos=0;
> +	while(length) {
> +
> +		size_t bytes_to_copy = frame_size-input_buffer_pos;
> +		if(bytes_to_copy > length) {
> +			bytes_to_copy = length;	
>
> Why not |std::min()|?
>
fixed.
> ------------------------------------------------------------------------
>
> In src/encoder/plugins/FdkaacEncoderPlugin.cxx 
> <https://github.com/MusicPlayerDaemon/MPD/pull/287#discussion_r193001113>:
>
> > +#include <fdk-aac/aacenc_lib.h>
> +
> +#include <stdexcept>
> +
> +#include <assert.h>
> +#include <string.h>
> +
> +class FdkaacEncoder final : public Encoder {
> +	const AudioFormat audio_format;
> +	HANDLE_AACENCODER *AacEncoder;
> +	AACENC_InfoStruct *info;
> +	int input_buffer_pos = 0;
> +
> +	ReusableArray<unsigned char, 32768> output_buffer;
> +	ReusableArray<unsigned char, 32768> input_buffer;
> +	unsigned char *output_begin = nullptr, *output_end = nullptr, *input_begin = nullptr;
>
> These attributes should be documented.
>
fixed
> ------------------------------------------------------------------------
>
> In src/encoder/plugins/FdkaacEncoderPlugin.cxx 
> <https://github.com/MusicPlayerDaemon/MPD/pull/287#discussion_r193001573>:
>
> > +
> +#include <fdk-aac/aacenc_lib.h>
> +
> +#include <stdexcept>
> +
> +#include <assert.h>
> +#include <string.h>
> +
> +class FdkaacEncoder final : public Encoder {
> +	const AudioFormat audio_format;
> +	HANDLE_AACENCODER *AacEncoder;
> +	AACENC_InfoStruct *info;
> +	int input_buffer_pos = 0;
> +
> +	ReusableArray<unsigned char, 32768> output_buffer;
> +	ReusableArray<unsigned char, 32768> input_buffer;
>
> Why |32768|? Do you know what this template parameter means?
>
This code has been taken from your lame plugin. 8192 could be enough as 
input buffer, and about 1024 bytes could be enough as output buffer, but 
this  does not really matter.

> ------------------------------------------------------------------------
>
> In src/encoder/plugins/FdkaacEncoderPlugin.cxx 
> <https://github.com/MusicPlayerDaemon/MPD/pull/287#discussion_r193001768>:
>
> > +#include "util/NumberParser.hxx"
> +#include "util/ReusableArray.hxx"
> +#include "util/RuntimeError.hxx"
> +
> +#include <fdk-aac/aacenc_lib.h>
> +
> +#include <stdexcept>
> +
> +#include <assert.h>
> +#include <string.h>
> +
> +class FdkaacEncoder final : public Encoder {
> +	const AudioFormat audio_format;
> +	HANDLE_AACENCODER *AacEncoder;
> +	AACENC_InfoStruct *info;
> +	int input_buffer_pos = 0;
>
> No signed integer! This is asking for trouble.
>
fixed.
> ------------------------------------------------------------------------
>
> In src/encoder/plugins/FdkaacEncoderPlugin.cxx 
> <https://github.com/MusicPlayerDaemon/MPD/pull/287#discussion_r193002495>:
>
> > +		}
> +
> +		const auto input_buf = input_buffer.Get(bytes_to_copy);
> +		if(input_begin == nullptr) {
> +			input_begin = input_buf;
> +		}
> +
> +		memcpy(input_buf + input_buffer_pos, src, bytes_to_copy);
> +
> +		input_buffer_pos += bytes_to_copy;
> +		read_buffer_pos += bytes_to_copy;
> +		length -= bytes_to_copy;
> +		src += bytes_to_copy;
> +
> +		if(input_buffer_pos == frame_size) {
> +			const auto dest = output_buffer.Get(info->maxOutBufBytes);
>
> Does this overwrite the contents of your output buffer in each loop 
> iteration?
>
yes, this code has been taken from your lame plugin.

> ------------------------------------------------------------------------
>
> In src/encoder/plugins/FdkaacEncoderPlugin.cxx 
> <https://github.com/MusicPlayerDaemon/MPD/pull/287#discussion_r193002588>:
>
> > +			in_buf.bufSizes = &in_size;
> +			in_buf.bufElSizes = &in_elem_size;
> +
> +			int out_identifier = OUT_BITSTREAM_DATA;
> +			void *out_ptr = dest;
> +			int out_size = info->maxOutBufBytes;
> +			int out_elem_size = 1;
> +			out_buf.numBufs = 1;
> +			out_buf.bufs = &out_ptr;
> +			out_buf.bufferIdentifiers = &out_identifier;
> +			out_buf.bufSizes = &out_size;
> +			out_buf.bufElSizes = &out_elem_size;
> +
> +			AACENC_ERROR res = aacEncEncode ( *AacEncoder, &in_buf, &out_buf, &in_args, &out_args );
> +			if(res != AACENC_OK) {
> +				if (res == AACENC_ENCODE_EOF) {
>
> Why this redundant comparison?
>
because there is more than one single error condition. Only one of them 
is critical.

> ------------------------------------------------------------------------
>
> In src/encoder/plugins/FdkaacEncoderPlugin.cxx 
> <https://github.com/MusicPlayerDaemon/MPD/pull/287#discussion_r193002691>:
>
> > +			}
> +
> +			input_begin = nullptr;
> +			input_buffer_pos = 0;
> +		}
> +	}
> +}
> +
> +size_t
> +FdkaacEncoder::Read(void *dest, size_t length)
> +{
> +	const auto begin = output_begin;
> +	assert(begin <= output_end);
> +	const size_t remainning = output_end - begin;
> +	if (length > remainning)
> +		length = remainning;
>
> |std::min()|
>
This code has been taken from your lame plugin.

baycom commented:
On 05.06.18 11:26, Max Kellermann wrote:
>
> Some research about Fraunhofer FDK AAC library showed that the library 
> is non-free, thus incompatible with MPD's license, and your code would 
> be illegal in MPD. Sorry.
>
Well, this is really sad. You should think about your license.

MaxKellermann commented:
No, Fraunhofer should think about their license.
